### DISCUSSION TITLE
The output results are inconsistent after each execution

### DISCUSSION THREAD

**Discussion by minhaics** [nodejs/help#4789] [Q&A]
### Version

v18.18.0

### Platform

x64 window

### What steps will reproduce the bug?

```

let sleep_cpu_bound = function (time) {
    return new Promise((resolve, reject) => {
        var stop = new Date().getTime();
        while (new Date().getTime() < stop + time) {
            ;
        }
   
        return resolve(true)
    });
}
 
async function call_api_1() {
    try {
        console.log('entering the fetchData function');
 
        await sleep_cpu_bound(1000);
  
 
        const response_1 = await fetch('https://jsonplaceholder.typicode.com/posts/1');
 
        if (!response_1.ok) {
            throw new Error(`Error when calling API_1: ${response_1.status}`);
        }
 
        const data = await response_1.json();
        console.log('API_1 is coming');
 
        return data;
    } catch (error) {
        console.error('Error when calling API 1:', error);
        throw error;
    }
}
 
async function call_api_2() {
    try {
        console.log('entering the fetchSlowData function');
 
        const response_2 = await fetch('https://jsonplaceholder.typicode.com/comments/1');
        if (!response_2.ok) {
            throw new Error(`Error when calling API_2: ${response_2.status}`);
        }
 
        const data = await response_2.json();
        console.log('API_2 is coming');
        return data;
    } catch (error) {
        console.error('Error when calling API 2:', error);
        throw error;
    }
}
 
function checkings_1() {
    console.log('checking_1')
}
 
function checkings_2() {
    console.log('checking_2')
}
 
function checkings_3() {
    console.log('checking_3')
}
 
async function main() {
    const promise2 = call_api_2();
    console.log('promise_2', promise2);
    const promise1 = call_api_1();
    console.log('promise_1', promise1);
 
    checkings_1();
    checkings_2();
 
    //const [data2, data1] = await Promise.all([promise2, promise1]); // Wait for both promises to complete
 
    checkings_3();
    console.log('Ending');
}
 
main()
```

### How often does it reproduce? Is there a required condition?

Occasionally, and inconsistently across executions.

### What is the expected behavior? Why is that the expected behavior?
entering the fetchSlowData function
promise_2 Promise { <pending> }
entering the fetchData function
promise_1 Promise { <pending> }
checking_1
checking_2
checking_3
Ending
API_2 is coming
API_1 is coming

### What do you see instead?

entering the fetchSlowData function
promise_2 Promise { <pending> }
entering the fetchData function
promise_1 Promise { <pending> }
checking_1
checking_2
checking_3
Ending
API_1 is coming
API_2 is coming

### Additional information

hi You, in this piece of code, I intentionally want to notify 'API_2 is coming' first, so I placed the 'sleep()' function inside the 'call_api_1' function. However, when I run it on Node v18, I receive a strange result where sometimes 'API_2 is coming' appears first, and other times 'API_1 is coming' appears first on the screen. But when I check it in the Chrome browser, I consistently get the result 'API_2 is coming' first. I've tried to explain it, but I don't know the reason why. It seems to be a Node.js issue.

thank you!

**Comment by bnoordhuis** [comment]
Converted to a discussion because that's not a bug but a misunderstanding on your part.

You don't `await` the results of the async function calls, you `Promise.all(...)` them. That makes them run in parallel and they therefore complete in non-deterministic order.

**Reply by minhaics** [reply]
hi, I intentionally didn't await async function calls to allow both functions to run concurrently, that was my intention. However, I used '`await sleep_cpu_bound(1000)`;' inside the 'call_api_1' function, meaning I expected the second API call to complete first and display on the screen first. For some reason, why am I getting inconsistent results? I've tried this on bun runtime 1.0.3 and Chrome browser version 117, and I always get a consistent output, which is 'API_2 is coming' always appearing first.
The line of code `const [data2, data1] = await Promise.all([promise2, promise1])`;  I believe does not affect the issue I'm reporting because even when I comment them out, I still get inconsistent results. Sometimes, the 'API_2 is coming' message appears first, and sometimes the 'API_1 is coming' message appears first.

**Reply by bnoordhuis** [reply]
Your sleep_cpu_bound blocks the entire program until that many milliseconds have expired, doesn't matter if it's a promise or not.

**Reply by minhaics** [reply]
This might not be entirely a bug but rather a behavior of Node.js. However, I'm blocking the event loop with the await sleep_cpu_bound() line, but before that, the program has already initiated the second API call. According to logic, after sleep_cpu_bound finishes, the API_2 call should have received a response while the first API call hasn't been made yet. So, why am I getting inconsistent results in this case? I've tried this on both Chrome and Bun runtime, and I consistently receive the 'API_2 is coming' message first. It's possible that my report might have issues due to my understanding of the event loop.  I hope to receive your understanding regarding this matter.

**Reply by bnoordhuis** [reply]
> after sleep_cpu_bound finishes, the API_2 call should have received a response

This is the flaw in your reasoning: it doesn't. Replace sleep_cpu_bound with sleep so the event loop can process the fetch request in the background.
```js
function sleep(ms) { return new Promise(k => setTimeout(k, ms)) }
```

**Reply by minhaics** [reply]
If this is not an error, I find it challenging to predict exactly what will be displayed on the screen when programming in the future. Once I received consistent results when testing on the browser and runtime, personally, I think the message 'API_2 is coming' appearing first would make sense because the API_2 call is made before the main thread is blocked."

**Reply by minhaics** [reply]
I know this function 'function sleep(ms) { return new Promise(k => setTimeout(k, ms)) }`' is effective when I want to sleep. However, I use the 'sleep_cpu_bound' function to illustrate briefly the blocking of the main thread. In programming in general, there can be any other CPU-bound workload, for example, a recursive algorithm, in many cases. The fact that I use 'sleep_cpu_bound' is just an example to illustrate concisely that the current inconsistency issue exists

**Reply by bnoordhuis** [reply]
Remember node is single-threaded. If you block the main thread, you also block network I/O. Chrome offloads network I/O to a thread pool.

Your code is still defective with chrome though. If that API endpoint is slow to reply, you get results in a different order.

Key takeaway: don't use Promise.all if you want predictable order.

**Reply by minhaics** [reply]
According to the Node.js philosophy of doing other work while waiting for I/O, you mentioned, "If you block the main thread, you also block network I/O." I still don't fully understand. Based on the code above, could you tell me at the moment the event loop is blocked by the 'sleep_cpu_bound' function, whether the second API call has already been initiated? Has the 'fetch' function been triggered beforehand? You seem to have a deep understanding of Node.js behavior, and I would appreciate your help. In the case where it has already been triggered, with a blocking time of 1000ms, meaning the API2  call  has already received a response, and it's just waiting for the event loop to display it on the screen '`const response_2 = await fetch('https://jsonplaceholder.typicode.com/comments/1')`;.

**Reply by minhaics** [reply]
you mentioned that "Chrome offloads network I/O to a thread pool." But when I tested it on Bun, I consistently saw the message 'API_2 is coming' displayed first.

**Reply by minhaics** [reply]
The announcement 'api_2 is coming' appearing first does indeed make sence in the code logic above as the API 1 call has been blocked by the sleep_cpu_bound function, and it holds significance in programming in general. It allows us to predict what the output will be, rather than leaving it uncertain which one will come first. I tested it on Bun, and it works consistency. I hope you can seriously consider whether this might be a bug in Node.js or not

**Reply by bnoordhuis** [reply]
It's not a node bug. I think you think the sleep_cpu_bound call holds up the first promise and not the other but that's where you're wrong: it blocks the whole program. Nothing happens in parallel when it's hogging the CPU.

Google the difference between "concurrent" and "parallel". Network I/O in node is concurrent but not parallel.

**Reply by minhaics** [reply]
hi @bnoordhuis  I still don't understand what you're saying. In my understanding, when sleep_cpu_bound blocks the entire program, it means it's consuming CPU. However, the first API call was already made and sent before, and that doesn't consume CPU. So why was it sent, and why is it being blocked? It might be blocked from receiving the response, but when the main thread is released, it will receive that response. Compared to the fact that the API_1 call hasn't been made yet, the event loop should receive a response from the API_2 call earlier. That makes sense because it optimizes time efficiently, However, once again, those are the outputs I receive in Bun and the browser, but with Node.js, I get unpredictable and inconsistent outputs.

**Reply by bnoordhuis** [reply]
I don't know if I can explain it to you because you keep making unfounded assumptions. For example, your intuition about when the request is sent and the response received has little basis in reality and tells me you haven't internalized the difference between concurrent and parallel.

I'm going to bow out of this discussion. I wish you luck on your endeavors.

**Reply by minhaics** [reply]
Thank you very much, Ben!
I am beginning to understand your points. Besides I/O networking being blocked entirely, including API calls like HTTP requests that are made  before the main thread is blocked, I have been wondering about file system APIs (such as fs.writeFile) and crypto. Do these functions  use the thread pool, and are they blocked when the main thread is being blocked, or do they run in the background even when the main thread is being blocked?  Both use a worker pool, but is there a difference in running in parallel with the main thread?

I truly appreciate any guidance or assistance you can provide.

**Reply by bnoordhuis** [reply]
> Do these functions use the thread pool

fs: yes, crypto: sometimes (ex. randomBytes but not f.e. encrypt/decrypt)

> do they run in the background even when the main thread is being blocked

Correct. I do believe you've internalized the difference between concurrent and parallel. :-)

**Reply by minhaics** [reply]
```
const fs = require('fs').promises;

// Function to generate large and heavy content
function generateHeavyContent(sizeInMB) {
    const content = 'This is some heavy content. '.repeat(10); // A line of content with a length of 1024 characters
    const iterations = Math.floor(sizeInMB * 1024 * 1024 / content.length); // Number of iterations to create the desired content size (MB * 1024 * 1024)
    return content.repeat(iterations);
}

// Asynchronous function to write heavy content to a file without blocking the main thread
async function writeHeavyFile(filePath, sizeInMB, index) {
    const heavyContent = generateHeavyContent(sizeInMB);
    console.log("starting timing");
    const startTime = process.hrtime();
    try {
        await fs.writeFile(filePath, heavyContent);
        const endTime = process.hrtime(startTime);
        const executionTimeInSeconds = endTime[0] + endTime[1] / 1e9; // Convert to seconds
        console.log('index', index);
        console.log(`Execution time: ${executionTimeInSeconds.toFixed(2)} seconds`);
        console.log(`File written successfully to ${filePath}`);
    } catch (error) {
        console.error(`Error writing file: ${error}`);
    }
}

sleepP = function (time) {
    return new Promise((resolve, reject) => {
        console.log('starting sleep')
        var stop = new Date().getTime();
        while (new Date().getTime() < stop + time) {
            ;
        }
        return resolve(true)
    });
}

// Using the function to write heavy content to files
const outputFile = 'heavy_file.txt'; // Output file name
const fileSizeInMB = 100; // Output file size (MB)

async function main() {
    for (let i = 0; i < 7; i++) {
        const outputFile = `heavy_file_${i}.txt`; // Using template literals to generate dynamic file names
        writeHeavyFile(outputFile, fileSizeInMB, i);
    }
    await sleepP(20000); // Sleep for 20 seconds
}

main();


//output:
// starting timing
// starting timing
// starting timing
// starting timing
// starting timing
// starting timing
// starting timing
// starting sleeP
// index 0
// Execution time: 29.98 seconds
// File written successfully to heavy_file_0.txt
// index 5
// Execution time: 29.28 seconds
// File written successfully to heavy_file_5.txt
// index 4
// Execution time: 29.43 seconds
// File written successfully to heavy_file_4.txt
// index 1
// Execution time: 29.85 seconds
// File written successfully to heavy_file_1.txt
// index 2
// Execution time: 29.74 seconds
// File written successfully to heavy_file_2.txt
// index 3
// Execution time: 30.22 seconds
// File written successfully to heavy_file_3.txt
// index 6
// Execution time: 29.81 seconds
// File written successfully to heavy_file_6.txt



```

**Reply by minhaics** [reply]
Dear friend @bnoordhuis, in this code snippet, I have made 7 writeHeavyFile functions run concurrently, and the main_thread is blocked by the sleepP function for a period of 20 seconds. This should be enough time for the writeHeavyFile functions to run in the background and complete file writing. However, I do not observe them running in the background as expected.

Even after the main_thread is unblocked, it seems that the 7 writeHeavyFile functions are still running on only one thread. I have tried setting "UV_THREADPOOL_SIZE=8" or "UV_THREADPOOL_SIZE=1," but the computation time remains the same.

This experiment suggests that the fs.writeFile(filePath, heavyContent) function does not use threads from the worker_pool. I hope you can help me clarify this issue. 

Previously, I asked you whether the file read/write functions from fs module run in the background when the main thread is blocked. You said they do, and you also mentioned that they use threads from the worker pool, However, what I experimented with didn't yield the results as you described. Could you please correct my example or provide me with an example that helps me better understand what you mentioned earlier?

Thank you

**Reply by bnoordhuis** [reply]
fs.writeFile() round-trips through the thread pool operations at least three times: open file, write data, close file. You block the event loop after the first operation.

**Reply by minhaics** [reply]
Dear @bnoordhuis, you mentioned three activities: open file, write data, close file. To initiate any of these activities, does it require activation from the main thread? If the main thread is blocked, can they not be invoked? Could you tell me if the 'write data' operation is atomic? Does it write in chunks by default? If it writes in chunks, does it still preserve atomicity?

I can't think of a code example where a program can write a file while the event loop is being blocked. As you can see in the code snippet above,  the worker pool only just 'opened the file' before 'writing data,' and I have tried to illustrate a code snippet that demonstrates that even though the event loop is blocked by a CPU-intensive task, file writing operations continue, but I can't find or think of any such code example.

**Reply by bnoordhuis** [reply]
> does it require activation from the main thread? If the main thread is blocked, can they not be invoked?

Yes and yes.

> the 'write data' operation is atomic?

I'm not sure what you mean by "atomic" here. In general partial data can be observed by other parties, because that's how most file systems work, they're not ACID databases.

**Comment by rehamdan** [comment]
https://www.ju.edu.jo/Home.aspx
